src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,766 @@
+import org.jclouds.io.ContentMetadata;
+import org.jclouds.io.Payload;
+import org.jclouds.io.Payloads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("UnstableApiUsage")
+public final class EncryptedBlobStore extends ForwardingBlobStore {
+    private final Logger logger =
+        LoggerFactory.getLogger(EncryptedBlobStore.class);
+    private SecretKeySpec secretKey;
+
+    private EncryptedBlobStore(BlobStore blobStore, Properties properties)
+        throws IOException {
+        super(blobStore);
+
+        String password = properties.getProperty(
+            S3ProxyConstants.PROPERTY_ENCRYPTED_BLOBSTORE_PASSWORD);
+        if (password == null || password.isEmpty()) {
+            throw new IOException(
+                "Password for encrypted blobstore is not set");
+        }
+        String salt = properties.getProperty(
+            S3ProxyConstants.PROPERTY_ENCRYPTED_BLOBSTORE_SALT);
+        if (salt == null || salt.isEmpty()) {
+            throw new IOException("Salt for encrypted blobstore is not set");
+        }
+        initStore(password, salt);
+    }
+
+    static BlobStore newEncryptedBlobStore(BlobStore blobStore,
+        Properties properties) throws IOException {
+        return new EncryptedBlobStore(blobStore, properties);
+    }
+
+    private void initStore(String password, String salt)
+        throws IllegalArgumentException {
+        try {
+            SecretKeyFactory factory =
+                SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
+            KeySpec spec =
+                new PBEKeySpec(password.toCharArray(), salt.getBytes(), 65536,
+                    128);
+            SecretKey tmp = factory.generateSecret(spec);
+            secretKey = new SecretKeySpec(tmp.getEncoded(), "AES");
+        } catch (Exception e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private Blob cipheredBlob(String container, Blob blob, InputStream payload,
+        long contentLength,
+        boolean addEncryptedMetadata) {
+
+        // make a copy of the blob with the new payload stream
+        BlobMetadata blobMeta = blob.getMetadata();
+        ContentMetadata contentMeta = blob.getMetadata().getContentMetadata();
+        Map<String, String> userMetadata = blobMeta.getUserMetadata();
+        String contentType = contentMeta.getContentType();
+
+        // suffix the content type with -s3enc if we need to encrypt
+        if (addEncryptedMetadata) {
+            blobMeta = setEncryptedSuffix(blobMeta);
+        } else {
+            // remove the -s3enc suffix while decrypting
+            // but not if it contains a multipart meta
+            if (!blobMeta.getUserMetadata()
+                .containsKey(Constants.METADATA_IS_ENCRYPTED_MULTIPART)) {
+                blobMeta = removeEncryptedSuffix(blobMeta);
+            }
+        }
+
+        // we do not set contentMD5 as it will not match due to the encryption
```

Comment:
I would like to document this shortcoming for now if it is ok for you?

Reply:
Honestly the Content-MD5 header is less valuable than you imagine -- I believe that the AWS SDK does not bother to calculate MD5 if either HTTPS or v4 X-Amz-Content-SHA256 are used since these provide sufficient transport-level security.  I don't think anything needs to be done here.